README.md
# Arduino CS5530

[![Build Status](https://travis-ci.org/sandeepmistry/arduino-LoRa.svg?branch=master)](https://travis-ci.org/sandeepmistry/arduino-LoRa)

An [Arduino](https://arduino.cc/) library for sending and receiving data using [Cirrus logic CS5530](https://www.lora-alliance.org/) .

## Compatible Hardware

 * [CS5530](https://www.cirrus.com/products/cs5530/) based boards including:
   * [CS5530 Shield v1.0](https://www.aliexpress.com/item/1005002054335619.html)
   * [CS5530 Sheild v2.0](https://www.aliexpress.com/item/1005002394037116.html) 
  

### Cirrus logic CS5530 wiring

| Cirrus logic CS5530 | Arduino |
| :---------------------: | :------:|
| VCC | 5.0V |
| GND | GND |
| SCK | SCK |
| SDO | MISO |
| SDI | MOSI |
| NSS | 10 |


`NSS` pin can be changed by using `cell.setPins(ss, reset, dio0)`. `DIO0` pin is optional, it is only needed for receive callback mode. If `DIO0` pin is used, it **must** be interrupt capable via [`attachInterrupt(...)`](https://www.arduino.cc/en/Reference/AttachInterrupt).

**NOTES**:
 * Some boards (like the Arduino Nano), cannot supply enough current for the CS5530 in TX mode. This will cause lockups when sending, be sure to use an external 3.3V supply that can provide at least 120mA's when using these boards.
 * If your Arduino board operates at 5V or 3.3V, no need to use a level converter for the wiring to the Cirrus logic CS5530 module. 

## Installation

### Using the Arduino IDE Library Manager

1. Choose `Sketch` -> `Include Library` -> `Manage Libraries...`
2. Type `CS5530` into the search box.
3. Click the row to select the library.
4. Click the `Install` button to install the library.

### Using Git

```sh
cd ~/Documents/Arduino/libraries/
git clone https://github.com/yasir-shahzad/CS5530 
```

## API

See [API.md](API.md).

## Examples

See [examples](examples) folder.

## FAQ

**1) Initializing the CS5530 is failing**

Please check the wiring you are using matches what's listed in [Cirrus logic CS5530](# Cirrus logic CS5530-wiring). You can also use `cell.setPins(ss)` to change the default pins used. You can call `cell.setSPIFrequency(frequency)` to lower the SPI frequency used by the library. Both API's must be called before `cell.begin(...)`.

**2) Is CS5530 MSBFIRST or LSBFIRST?**

If we see in the data sheet we can clearly see the CS5530 is MSBFIRST.

**3) Which SPI mode does CS5530 uses?**

if we see at Cpoles and Cphases in data-sheet, we can clearly analize from graphs that it's SPI_MODE0.

**4) Does it SPI works on fixed frequency?**

No, It's SPI work on different frequency ranges.


## License

This libary is [licensed](LICENSE) under the [MIT Licence](https://en.wikipedia.org/wiki/MIT_License).


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## [LHE-0.1] - 2022-02-28
### Added 
- [PD-2066] Skip confirmation

src/CS5530.h
/*
 * Copyright (c) 2022 by Yasir Shahzad <Yasirshahzad918@gmail.com>
 *
 * This file is free software; you can redistribute it and/or modify
 * it under the terms of either the GNU General Public License version 2
 * or the GNU Lesser General Public License version 2.1, both as
 * published by the Free Software Foundation.
 */


#ifndef CS5530_H
#define CS5530_H


//Alias of Basic Types
#define i32 long int
#define i16 int
#define i8  char
#define u32 unsigned long int
#define u16 unsigned int
#define u8  unsigned char


#define RESET_TIME 1200000
#define WASTE_TIME 90000000


#define SELF_OFFSET     1
#define SELF_GAIN       2
#define SYSTEM_OFFSET   3
#define SYSTEM_GAIN     4

#define SINGLE_CONVERSION       1
#define CONTINUED_CONVERSION    2


//Alias of Basic Results
#define PASS              0x0
#define FAIL              0xff
#define TRUE              1
#define FALSE             0
 
#define NOTHING 0

#define CS5530_READ_DATA_TIMEOUT 2000000

#define TIMEOUTERR         0xff
 
#define SUCCESSFUL         0x0

#define CMD_STOP_CONT_CONV 0xff

#define CAL_TIMEOUT_LIMIT  4000

#define READ_CONV_RESULT   0x00


#define DATA_VALID  0
 
#define ERR_AD_BUSY 1
 
#define ERR_AD_OVER_FLOW   2
 
#define ERR_FILTER_ONGOING 3


//2.3.9. Configuration Register Descriptions
#define REG_CONFIG_PSS      1UL << 31     //Power Save Select
#define REG_CONFIG_PDM      1UL << 30     //Power Down Mode
#define REG_CONFIG_RS       1UL << 29     //Reset System
#define REG_CONFIG_RV       1UL << 28     //Reset Valid
#define REG_CONFIG_IS       1UL << 27     //Input Short
#define REG_CONFIG_VRS      1UL << 25     //Voltage Reference Select
#define REG_CONFIG_A1       1UL << 24
#define REG_CONFIG_A0       1UL << 23
#define REG_CONFIG_FRS      1UL << 19     //Filter Rate Select
#define REG_CONFIG_OCD      1UL << 9

//CS5532 Gain settings
#define GAINX1              1UL  << 24
#define GAINX2              2UL  << 24
#define GAINX4              4UL  << 24
#define GAINX8              8UL  << 24
#define GAINX16             16UL << 24
#define GAINX32             32UL << 24
#define GAINX64             64UL << 24

//Word Rate when FSR=1
#define WORD_RATE_100SPS    0x0 << 11
#define WORD_RATE_50SPS     0x1 << 11
#define WORD_RATE_25SPS     0x2 << 11
#define WORD_RATE_12P5SPS   0x3 << 11
#define WORD_RATE_6P25SPS   0x4 << 11
#define WORD_RATE_3200SPS   0x8 << 11
#define WORD_RATE_1600SPS   0x9 << 11
#define WORD_RATE_800SPS    0xA << 11
#define WORD_RATE_400SPS    0xB << 11
#define WORD_RATE_200SPS    0xC << 11

//Unipolar / Bipolar
#define CS5530_UNIPOLAR     1UL << 10
#define CS5530_BIPOLAR      0UL << 10
#define REG_DATA_OF         1UL << 3

#define CMD_STOP_CONT_CONV  0xFF

#define 5530_DEFAULT_SPI           SPI
#define 5530_DEFAULT_SPI_FREQUENCY 8E6 
#define 5530_DEFAULT_SS_PIN        10



//Command Register (CS5530_COMM_REG)
#define CMD_OFFSET_READ         0x09
#define CMD_OFFSET_WRITE        0x01
#define CMD_GAIN_READ           0x0A
#define CMD_GAIN_WRITE          0x02
#define CMD_CONFIG_READ         0x0B
#define CMD_CONFIG_WRITE        0x03

#define CMD_CONVERSION_SIGLE    0x80
#define CMD_CONVERSION_CONTINU  0xC0
#define CMD_SYS_OFFSET_CALI     0x85
#define CMD_SYS_GAIN_CALI       0x86
#define CMD_SYNC1               0xFF          //Part of the serial port re-initialization sequence.
#define CMD_SYNC0               0xFE          //End of the serial port re-initialization sequence.
#define CMD_NULL                0x00          //This command is used to clear a port flag and keep the converter in the continuous conversion mode.


enum EAdStatus {
   E_AD_STATUS_BUSY,
   E_AD_STATUS_READY,
   E_AD_STATUS_OVERFLOW
};



class CS5530 {
public:
	
 CS5530();
 int begin();
 void setPin(int ss = 5530_DEFAULT_SS_PIN);
 void setSPI(SPIClass& spi);
 void setSPIFrequency(u32);
 u32 twoComplement(u32);
 u8 readByte(void);
 u32 read4Bytes(void);
 u32 readRegister(u8);
 u8 convert(u8, u8, u8, int);
 u8 calibrate(u8, int, int);
 void write4Bytes(u32);
 void writeByte(u8);
 void writeRegister(u8, u32);
 void setBit(u8, u32);
 void resetBit(u8, u32);
 bool isReady(void);
 bool reset(void);
 u32  readWeightsclae();


private:
 SPISettings _spiSettings;
 SPIClass* _spi;
 int _ss;
};

#endif






src/CS5530.cpp
#include <string.h>
#include <avr/pgmspace.h>
#include <SPI.h>
#include "CS5530.h"
#include "Arduino.h"
#include <stdio.h>


CS5530::CS5530():
_spiSettings(5530_DEFAULT_SPI_FREQUENCY, MSBFIRST, SPI_MODE0),
_spi(&5530_DEFAULT_SPI), _ss(5530_DEFAULT_SS_PIN)

{
  _spi->begin();
}

int CS5530::begin()
{
  // setup pins
  pinMode(_ss, OUTPUT);

   digitalWrite(_ss, LOW);
   
   
  // set SS high
  digitalWrite(_ss, HIGH);

  // start SPI
  _spi->begin();

  return 1;
}


void CS5530::setPin(int ss)
{
  _ss = ss;
}

void CS5530::setSPI(SPIClass& spi)
{
  _spi = &spi;
}

void CS5530::setSPIFrequency(u32 frequency)
{
  _spiSettings = SPISettings(frequency, MSBFIRST, SPI_MODE0);
}

bool CS5530::reset(void) {
    int i;
    u32 tmp;
	
    //Initilizing SPI port
    for(i=0;i<15;i++) {
      writeChar(CMD_SYNC1);
    }
	
    writeChar(CMD_SYNC0);
	
    // Reseting CS5530
    writeRegister(CMD_CONFIG_WRITE,  REG_CONFIG_RS);    	
    delay(1);  //Wait 1 milli seconds
    writeRegister(CMD_CONFIG_WRITE, CMD_NULL); 
	
    tmp = readRegister(CMD_CONFIG_READ);	

    if(tmp & REG_CONFIG_RV) {
     return true;
    }

  return false;
}


void CS5530::writeRegister(u8 reg, u32 dat) {
 
    writeChar(reg);
    writeLong(dat);
}

void CS5530::setBit(u8 reg, u32 dat) {
    u32 tmp = 0;
    u8 cmd = 0;
    switch (reg)
    {
        case CMD_GAIN_WRITE:   cmd = CMD_GAIN_READ; break; 
        case CMD_OFFSET_WRITE: cmd = CMD_OFFSET_READ; break;		
        case CMD_CONFIG_WRITE: cmd = CMD_CONFIG_READ; break; 
    }

    tmp =  readRegister(cmd);
    tmp |= dat;
    writeChar(reg);
    writeLong(tmp);
}

void CS5530::resetBit(u8 reg, u32 dat) {
     u32 tmp = 0;
     u8 cmd = 0;
    switch (reg)
    {
	    case CMD_GAIN_WRITE:   cmd = CMD_GAIN_READ; break; 
	    case CMD_OFFSET_WRITE: cmd = CMD_OFFSET_READ; break;		
	    case CMD_CONFIG_WRITE: cmd = CMD_CONFIG_READ; break; 
    }

    tmp =  readRegister(cmd);
    tmp &= ~dat;
    writeChar(reg);
    writeLong(tmp);
}


void CS5530::writeChar(u8 dat) {
 
    digitalWrite(_ss, LOW);
	_spi->beginTransaction(_spiSettings);
    SPI.transfer(dat & 0xFF); 
	_spi->endTransaction();
    digitalWrite(_ss, HIGH);
}

void CS5530::writeLong(u32 dat) {
    int i;
    u8 tmp;

    for(i=3; i>=0; i--) {
        tmp = (u8)( (dat >> (8*i)) & 0xff);
        writeChar(tmp);
    }
}
 

u32 CS5530::readRegister(u8 reg) {
    u32 dat;
	
    writeChar(reg);
    dat = readLong();
	
    return dat;
}


u32 CS5530::readLong(void)      {
    int i;
    u32 dat=0; 
    u8 currntByte = 0;
   
    for(i=0; i<4; i++) {
        dat    <<=    8;
        dat    |= readChar();
    }

    return dat;
}

u8 CS5530::readChar(void)     {
    u8 dat=0;
	  
    digitalWrite(_ss, LOW);
	_spi->beginTransaction(_spiSettings);
    dat = SPI.transfer(CMD_NULL);
	_spi->endTransaction();
    digitalWrite(_ss, HIGH);
	  
    return dat;
}



bool CS5530::isReady(void) {
    if(digitalRead(12) == 0)
    {
        return true;
    }

    return false;
}




u32 CS5530::readWeightsclae() {
    u32 rec_data = 0;
    
    EAdStatus status;

   if(isReady()==false)
   {
       status = E_AD_STATUS_BUSY;
       return -2;
   }
   else
   {    
       rec_data = readRegister(CMD_NULL);
       if((rec_data &  REG_DATA_OF) == 0)
       {
           rec_data &= 0xFFFFFF00;
           
           rec_data = rec_data >> 8;

           status =  E_AD_STATUS_READY;

           return rec_data;
       }
       else
       {
             status =  E_AD_STATUS_OVERFLOW;
             return -1;
       }
   }

   // return status;
}


u8 CS5530::calibrate(u8 calibrate_type, int cfg_reg, int setup_reg) { 
	u32 calibrate_result;
	int waste_time, i;
	cfg_reg = (int)((calibrate_type % 2 == 1) ? (cfg_reg|REG_CONFIG_IS):(cfg_reg));
	u8 cmd,read_reg;

	writeRegister(CMD_CONFIG_WRITE, cfg_reg);
	writeChar(cmd);

	for(waste_time = WASTE_TIME; waste_time > 0; waste_time--)
		;

	calibrate_result = readRegister(read_reg);

	printf("The calibration result is: ");
	printf("%x ", calibrate_result);
	printf("\n");

	return 1; 
}

u32 CS5530::twoComplement(u32 n) {
    u32 negative = (n & (1UL << 23)) != 0;
	u32 native_int;

	if (negative)
	  native_int = n | ~((1UL << 24) - 1);
	else
	  native_int = n;
	return native_int;
}



u8 CS5530::convert(u8 convert_type, u8 setup_reg_no, u8 reg_no, int word_rate) {
	Serial.print("Prepare for conversion.\n");
	u32 final_result = 0;
	int waste_time, i;
	u32 cfg_reg =  (u32)REG_CONFIG_VRS;
	int setup_reg = ((setup_reg_no % 2 == 1) ? ((word_rate) << 16) : word_rate);
	u8 cmd;

	switch (convert_type)
	{
		case SINGLE_CONVERSION: cmd = CMD_CONVERSION_SIGLE; break;
		case CONTINUED_CONVERSION: cmd = CMD_CONVERSION_CONTINU; break; 
	}

	writeRegister(CMD_CONFIG_WRITE, cfg_reg);
    delay(10);
	//writeChar(cmd);
	Serial.print("Conversion begins...\n");

	delay(10);
	u8 test = 0;
//	test = readChar();   // wastercycles
	//final_result = readLong();


	Serial.print("The raw result is:"); Serial.println(final_result,BIN);
	Serial.print("The raw result is:"); Serial.println(final_result);

	final_result = twoComplement(final_result);
	Serial.print("The final result is:"); Serial.println(final_result,BIN);
	Serial.print("The fianl result is:"); Serial.println(final_result);

	final_result = final_result * 500 / 0x7fffff;
	Serial.print("The fianl result is:"); Serial.println(final_result);

	return 1;
}



